docs/source/meta_init.rst

```diff
@@ -0,0 +1,239 @@
+    :nosignatures:
+
+    meta_init
+    is_meta_init
+    materialize
+    clear_meta_init_cache
+
+.. autofunction:: meta_init
+.. autofunction:: is_meta_init
+.. autofunction:: materialize
+.. autofunction:: clear_meta_init_cache
+
+.. _common-failures:
+
+Common Failure Cases
+--------------------
+Since meta initialization relies on constructing modules and their tensors on
+the meta device, there are some certain usage patterns that will fail when used
+with :func:`meta_init`.
+
+**A module that uses an operator not supported by the meta backend:**
+If the module uses an operator that is not yet supported by the meta backend,
+the operator call will fail. If you experience such failure, please open a
+GitHub issue and let us know which operator you are using.
+
+**A module that has device-specific initialization logic:**
+A module that expects to be initialized using one or more specific device types
+and includes initialization logic such as:
+
+::
+
+  def MyModule(Module):
+      def __init__(self):
+          x = torch.ones([10, 10])
+          if x.is_cpu():
+              y = x + 1
+          elif x.is_cuda():
+              y = x + 2
+
+
+will silently fail, potentially with some partial initialization.
+
+**A module that uses auxiliary state beyond its parameters and buffers:**
+The :func:`materialize` function simply traverses through the parameters and
+buffers of a module and materializes them. This means if the module has some
+other auxiliary state those tensors won't be materialized.
+
+**A module that directly accesses to raw tensor storage:**
+If the module attempts to directly read from or write to the raw storage of a
+tensor using a naked pointer without checking its validity, it will cause a
+segmentation fault since meta tensors don't have any storage allocated to them.
+
+**A tensor loaded from external data:**
+Although not necessarily a failure, if a tensor is constructed from external
+data (e.g. numpy), :func:`meta_init` won't intercept it and will allow the
+tensor to be constructed with the actual data. Since the data is externally
+owned and already loaded into memory, this is beyond :func:`meta_init`'s
+control.
+
+Mitigation Strategy
+^^^^^^^^^^^^^^^^^^^
+If you are the module author, you can use the :func:`is_meta_init()`
+function to find out whether your module is being used within the scope of a
+:func:`meta_init` context and have an alternate logic if necessary:
+
+::
+
+    class MyModule(Module):
+        def __init__(self):
+            super().__init__()
+
+            if torch.nn.utils.is_meta_init():
+                self.my_buffer = torch.empty([10,10], device="meta")
+            else:
+                self.my_buffer = load_my_buffer()
+
+        # A function that does not support the meta backend.
+        def load_my_buffer(self) -> Tensor:
+            ...
+
+
+Alternative Approach
+--------------------
+If you are a module author, an alternative to :func:`meta_init()` is to have an
```

Comment:
did we figure out any way to compare these approaches? I.e. allow module customize its behavior but still be part of the larger tree which might not have explicit device args

Reply:
I am thinking of introducing three related features that I will mention in the design doc:

- A `no_meta_init` context manager that can be used inside a meta-init context similar to `torch.no_grad`, `torch.enable_grad`.
- A `__torch_materialize__` dunder method for Tensor subclasses implemented in Python to customize their behavior.